flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -493,18 +500,20 @@ void onSenderBacklog(int backlog) throws IOException {
 		int numRequestedBuffers = 0;
 
 		synchronized (bufferQueue) {
-			// Important: check the isReleased state inside synchronized block, so there is no
-			// race condition when onSenderBacklog and releaseAllResources running in parallel.
+			// Similar to notifyBufferAvailable(), make sure that we never add a buffer
+			// after releaseAllResources() released all buffers (see above for details).
 			if (isReleased.get() || isBlocked()) {
 				return;
 			}
 
 			numRequiredBuffers = backlog + initialCredit;
-			while (bufferQueue.getAvailableBufferSize() < numRequiredBuffers && !isWaitingForFloatingBuffers) {
+			int numAvailableBuffers = bufferQueue.getAvailableBufferSize();
```

Comment:
Please also revert this change. It can not have any performance effect, while it inlines some of the `bufferQueue` logic here.